packages/core/src/utils/validators.ts

```diff
@@ -0,0 +1,153 @@
+   * and the form control value as arguments and returns a
+   * boolean to evaluate for that Validator.
+   * @param instance {FormControlInterface} - The FormControl instance
+   * @param value {ValidationValue} - The form control value
+   * @returns {boolean} - The validity of a given Validator
+   */
+  isValid(instance: HTMLElement, value: ValidationValue): boolean;
+}
+
+export interface AsyncValidator extends ValidatorBase {
+  /**
+   * Callback for a given validator. Takes the FormControl instance
+   * and the form control value as arguments and returns a
+   * boolean to evaluate for that Validator as a promise.
+   * @param instance {FormControlInterface} - The FormControl instance
+   * @param value {ValidationValue} - The form control value
+   * @returns {Promise<boolean>} - The validity of a given Validator
+   */
+  isValid(instance: HTMLElement, value: ValidationValue, signal: AbortSignal): Promise<boolean | void>;
+}
+
+export type Validator = SyncValidator | AsyncValidator;
+
+export const requiredValidator: Validator = {
+  attribute: 'required',
+  key: 'valueMissing',
+  message: 'Please fill out this field',
+  isValid(instance: HTMLElement & { required: boolean }, value: ValidationValue): boolean {
+    let valid = true;
+
+    if ((instance.hasAttribute('required') || instance.required) && !value) {
+      valid = false;
+    }
+
+    return valid;
+  }
+};
+
+export const programmaticValidator: Validator = {
+  attribute: 'error',
+  message(instance: HTMLElement & { error: string }): string {
+    return instance.error;
+  },
+  isValid(instance: HTMLElement & { error: string }): boolean {
+    return !instance.error;
+  }
+};
+
+export const minLengthValidator: Validator = {
+  attribute: 'minlength',
+  key: 'rangeUnderflow',
+  message(instance: { minLength: number }, value: ValidationValue): string {
+    const _value = (value as string) || '';
+    return `Please use at least ${instance.minLength} characters (you are currently using ${_value.length} characters).`;
+  },
+  isValid(instance: HTMLElement & { minLength: number }, value: string): boolean {
+    /** If no value is provided, this validator should return true */
+    if (!value) {
+      return true;
+    }
+
+    if (!!value && instance.minLength > value.length) {
+      return false;
+    }
+
+    return true;
+  }
+};
+
+export const maxLengthValidator: Validator = {
+  attribute: 'maxlength',
+  key: 'rangeOverflow',
+  message(instance: { maxLength: number }, value: ValidationValue): string {
+    const _value = (value as string) || '';
+    return `Please use no more than ${instance.maxLength} characters (you are currently using ${_value.length} characters).`;
+  },
+  isValid(instance: HTMLElement & { maxLength: number }, value: string): boolean {
+    /** If maxLength isn't set, this is valid */
+    if (!instance.maxLength) {
+      return true;
+    }
+
+    if (!!value && instance.maxLength < value.length) {
```

Comment:
Here we can also simplify it to `return !(!!value && instance.maxLength < value.length);`?

Reply:
I would like to change this only after there are unit tests for the validators.